akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -95,14 +95,15 @@ object GraphStageLogic {
    */
   class ConditionalTerminateInput(predicate: () ⇒ Boolean) extends InHandler {
     override def onPush(): Unit = ()
-    override def onUpstreamFinish(): Unit = if (predicate()) inOwnerStageLogic.completeStage()
+    override def onUpstreamFinish(): Unit =
```

Comment:
@rkuhn The nested classes approach seems to work just fine: https://github.com/akka/akka/pull/18862